File: exercism/python/allergies/allergies.py


class Allergies(object):

    ALLERGENS = (
        'eggs',
        'peanuts',
        'shellfish',
        'strawberries',
        'tomatoes',
        'chocolate',
        'pollen',
        'cats'
    )

    def __init__(self, score):
        self._allergens = set(
            a for i, a in enumerate(self.ALLERGENS)
            if 1 << i & score > 0
        )

    def is_allergic_to(self, allergen):
        return allergen in self._allergens

    @property
    def lst(self):
        return list(self._allergens)

File: exercism/csharp/allergies/Allergies.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Allergies
{
    private List<string> allergens;
    private int score;

    public Allergies(int score)
    {
        this.allergens = new List<string> {
          "eggs",
          "peanuts",
          "shellfish",
          "strawberries",
          "tomatoes",
          "chocolate",
          "pollen",
          "cats"
        };
        this.score = score;
    }

    public bool AllergicTo (string allergen)
    {
        return (1 << allergens.IndexOf(allergen) & score) > 0;
    }

    public IEnumerable<string> List ()
    {
        return allergens.Where((_, i) => (1 << i & score) > 0);
    }
}
